File: channel_manager.py
import discord
import logging
from config import Config

logger = logging.getLogger('discord_bot')

class ChannelManager:
    def __init__(self):
        self.temp_channels = set()
        
    async def create_temporary_channel(self, guild, member):
        """Create a temporary voice channel"""
        try:
            channel_name = f"{Config.TEMP_CHANNEL_PREFIX}{member.name}'s Channel"
            
            # Create channel in category if configured, otherwise in guild root
            if Config.CATEGORY_ID:
                category = discord.utils.get(guild.categories, id=Config.CATEGORY_ID)
                channel = await guild.create_voice_channel(
                    name=channel_name,
                    category=category
                )
            else:
                channel = await guild.create_voice_channel(name=channel_name)
            
            # Set permissions for the channel creator
            await channel.set_permissions(member, 
                connect=True,
                speak=True,
                manage_channels=True,
                move_members=True
            )
            
            self.temp_channels.add(channel.id)
            return channel
            
        except discord.errors.Forbidden:
            logger.error("Bot doesn't have permission to create channels")
            return None
        except Exception as e:
            logger.error(f"Error creating temporary channel: {str(e)}")
            return None
            
    async def delete_temporary_channel(self, channel):
        """Delete a temporary voice channel"""
        try:
            if channel.id in self.temp_channels:
                await channel.delete()
                self.temp_channels.remove(channel.id)
                return True
        except discord.errors.NotFound:
            # Channel already deleted
            self.temp_channels.remove(channel.id)
        except Exception as e:
            logger.error(f"Error deleting temporary channel: {str(e)}")
        return False
        
    async def cleanup_orphaned_channels(self, guild):
        """Clean up any temporary channels that might have been left from a previous session"""
        try:
            for channel in guild.voice_channels:
                if (channel.name.startswith(Config.TEMP_CHANNEL_PREFIX) and 
                    len(channel.members) == 0):
                    await channel.delete()
                    logger.info(f"Cleaned up orphaned channel: {channel.name}")
        except Exception as e:
            logger.error(f"Error cleaning up orphaned channels: {str(e)}")

File: channelManager.js
const Config = require('./config');

class ChannelManager {
    constructor() {
        this.tempChannels = new Set();
    }

    async createTemporaryChannel(guild, member) {
        try {
            const channelName = `${Config.TEMP_CHANNEL_PREFIX}${member.user.username}'s Channel`;

            let channelOptions = {
                name: channelName,
                type: 2, // Voice channel type
                permissionOverwrites: [
                    {
                        id: member.id,
                        allow: ['Connect', 'Speak', 'ManageChannels', 'MoveMembers']
                    },
                    {
                        id: guild.id,
                        allow: ['Connect', 'Speak']
                    }
                ]
            };

            // Create in category if configured
            if (Config.CATEGORY_ID) {
                const category = guild.channels.cache.get(Config.CATEGORY_ID);
                if (category) {
                    channelOptions.parent = category.id;
                }
            }

            const channel = await guild.channels.create(channelOptions);

            this.tempChannels.add(channel.id);
            console.log(`Created temporary channel ${channel.name} for ${member.user.username}`);
            return channel;
        } catch (error) {
            console.error('Error creating temporary channel:', error);
            return null;
        }
    }

    async deleteTemporaryChannel(channel) {
        try {
            if (this.tempChannels.has(channel.id)) {
                await channel.delete();
                this.tempChannels.delete(channel.id);
                console.log(`Deleted temporary channel ${channel.name}`);
                return true;
            }
            return false;
        } catch (error) {
            if (error.code === 10003) { // Unknown Channel error
                this.tempChannels.delete(channel.id);
                return true;
            }
            console.error('Error deleting temporary channel:', error);
            return false;
        }
    }

    async cleanupOrphanedChannels(guild) {
        try {
            const channels = guild.channels.cache.filter(channel =>
                channel.type === 2 && // Voice channel
                channel.name.startsWith(Config.TEMP_CHANNEL_PREFIX) &&
                channel.members.size === 0
            );

            for (const [, channel] of channels) {
                await channel.delete();
                console.log(`Cleaned up orphaned channel: ${channel.name}`);
            }
        } catch (error) {
            console.error('Error cleaning up orphaned channels:', error);
        }
    }
}

module.exports = ChannelManager;